src/main/java/hudson/plugins/gradle/injection/CpPatternResource.java

```diff
@@ -0,0 +1,72 @@
+package hudson.plugins.gradle.injection;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Optional;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
+
+public class CpPatternResource {
+
+    private final Pattern filenamePattern;
+    private String resolved;
+
+    public CpPatternResource(Pattern filenamePattern) {
+        this.filenamePattern = filenamePattern;
+    }
+
+    public String resolve() {
+        if (resolved == null) {
+            resolved = findResource();
+        }
+        return resolved;
+    }
+
+    private String findResource() {
+        String notFound = "Could not find resource with filename pattern: " + filenamePattern;
+        try {
+            String directoryName = "hudson/plugins/gradle/injection/";
+            Enumeration<URL> resources = CopyUtil.class.getClassLoader().getResources(directoryName);
+            while (resources.hasMoreElements()) {
+                URL url = resources.nextElement();
+                if (url.getProtocol().equals("file")) {
+                    Optional<String> found = findInDir(url);
+                    if (found.isPresent()) {
+                        return found.get();
+                    }
+                } else if (url.getProtocol().equals("jar")) {
+                    Optional<String> found = findInJar(directoryName, url);
+                    if (found.isPresent()) {
+                        return found.get();
+                    }
+                }
+            }
+        } catch (URISyntaxException | IOException e) {
+            throw new IllegalStateException(notFound, e);
+        }
+        throw new IllegalStateException(notFound);
+    }
+
+    private Optional<String> findInJar(String directoryName, URL url) throws IOException {
+        String path = url.getPath();
+        String jarPath = path.substring(5, path.indexOf("!"));
```

Comment:
My understanding is that we want the jar file name without protocol and internal resource.
I would extract this to a method with a meaningful name to clarify the intent?

Reply:
yeah it basically transforms from `file:/path/to/some.jar!some/dir` to `/path/to/some.jar`

Reply:
I've extracted a method and created a simple test for doc